### ISSUE TITLE
Gen 2 - Amplify Outputs type issues after adding groups

### ISSUE THREAD

**Issue by Drillan767** [aws-amplify/amplify-js#14045]
### Environment information

```plain text
System:
  OS: macOS 15.1.1
  CPU: (8) arm64 Apple M1
  Memory: 161.67 MB / 8.00 GB
  Shell: /bin/zsh
Binaries:
  Node: 20.15.1 - ~/.nvm/versions/node/v20.15.1/bin/node
  Yarn: 1.22.22 - ~/.nvm/versions/node/v20.15.1/bin/yarn
  npm: 10.8.3 - ~/.nvm/versions/node/v20.15.1/bin/npm
  pnpm: undefined - undefined
NPM Packages:
  @aws-amplify/auth-construct: 1.5.0
  @aws-amplify/backend: 1.8.0
  @aws-amplify/backend-auth: 1.4.1
  @aws-amplify/backend-cli: 1.4.2
  @aws-amplify/backend-data: 1.2.1
  @aws-amplify/backend-deployer: 1.1.10
  @aws-amplify/backend-function: 1.8.0
  @aws-amplify/backend-output-schemas: 1.4.0
  @aws-amplify/backend-output-storage: 1.1.3
  @aws-amplify/backend-secret: 1.1.5
  @aws-amplify/backend-storage: 1.2.3
  @aws-amplify/cli-core: 1.2.0
  @aws-amplify/client-config: 1.5.2
  @aws-amplify/deployed-backend-client: 1.4.2
  @aws-amplify/form-generator: 1.0.3
  @aws-amplify/model-generator: 1.0.9
  @aws-amplify/platform-core: 1.2.2
  @aws-amplify/plugin-types: 1.5.0
  @aws-amplify/sandbox: 1.2.6
  @aws-amplify/schema-generator: 1.2.5
  aws-amplify: 6.10.0
  aws-cdk: 2.171.1
  aws-cdk-lib: 2.171.1
  typescript: 5.6.3
No AWS environment variables
No CDK environment variables
```

### Describe the bug

Hello,

This is my first ever AWS project, so I'm not exactly certain this is a real issue, even though I can't find a solution to this problem.

I'm building a small website with both an admin and a user dashboard, so I decided to add groups. So I followed [this documentation](https://docs.amplify.aws/react/build-a-backend/auth/concepts/user-groups/) and added 2 groups, 'ADMIN' and 'USER' to my `amplify/auth/resource.ts` file like so:

```ts
export const auth = defineAuth({
    loginWith: {
        email: { ... },
    },
    groups: ['ADMIN', 'USER'],
    triggers: {
        postConfirmation,
    },
    access: allow => [
        allow.resource(postConfirmation).to(['addUserToGroup']),
    ],
    userAttributes: {
        'custom:role': {
            dataType: 'String',
            mutable: true,
        },
    },
})
```

I also followed [this documentation](https://docs.amplify.aws/react/build-a-backend/functions/examples/add-user-to-group/) to automate the group adding on registration process. It works well, I just set 'USER' as the group a user will join.

But when trying to build the project, I got the following errors:

```
Argument of type '{ auth: { user_pool_id: string; aws_region: string; user_pool_client_id: string; identity_pool_id: string; mfa_methods: never[]; standard_required_attributes: string[]; username_attributes: string[]; ... 4 more ...; unauthenticated_identities_enabled: boolean; }; data: { ...; }; storage: { ...; }; version: string; }' is not assignable to parameter of type 'ResourcesConfig | LegacyConfig | AmplifyOutputs'.
  Type '{ auth: { user_pool_id: string; aws_region: string; user_pool_client_id: string; identity_pool_id: string; mfa_methods: never[]; standard_required_attributes: string[]; username_attributes: string[]; ... 4 more ...; unauthenticated_identities_enabled: boolean; }; data: { ...; }; storage: { ...; }; version: string; }' is not assignable to type 'AmplifyOutputs'.
    The types of 'auth.groups' are incompatible between these types.
      Type '({ ADMIN: { precedence: number; }; USER?: undefined; } | { USER: { precedence: number; }; ADMIN?: undefined; })[]' is not assignable to type 'Record<string, UserGroupPrecedence>[]'.
        Type '{ ADMIN: { precedence: number; }; USER?: undefined; } | { USER: { precedence: number; }; ADMIN?: undefined; }' is not assignable to type 'Record<string, UserGroupPrecedence>'.
          Type '{ ADMIN: { precedence: number; }; USER?: undefined; }' is not assignable to type 'Record<string, UserGroupPrecedence>'.
            Property '"USER"' is incompatible with index signature.
              Type 'undefined' is not assignable to type 'UserGroupPrecedence'.

15 Amplify.configure(outputs)
```

The relevant part of the auto generated `amplify_outputs` looks like so:

```json
{
  "auth": {
    // ...
    "groups": [
      {
        "ADMIN": {
          "precedence": 0
        }
      },
      {
        "USER": {
          "precedence": 1
        }
      }
    ]
  }
}
```

Is there anything I didn't do correctly or at all? The auto generated output seems good to me.

Thank you in advance

### Reproduction steps

"Simply" add the following to the `amplify/auth/resource.ts` file:

```ts
groups: ['ADMIN' 'USER'],
```

And wait for the sandbox to regenerate.

**Comment by HuiSF** [comment]
Thanks for opening this issue. I was able to reproduce the same error thrown by Typescript. It looks like TS infers the type of the following JSON object differently than what the library expects:  

```ts
[
  {
    "ADMIN": {
      "precedence": 0
    }
  },
  {
    "USER": {
      "precedence": 1
    }
  }
]
// type =>
({
  ADMIN: { precedence: number; };
  USER?: undefined;
} | {
  USER: { precedence: number; };
  ADMIN?: undefined;
})[]
```

**Comment by HuiSF** [comment]
Hi @Drillan767 sorry for the inconvenience, if you are testing only the Auth category features, could you temporarily test with this version `aws-amplify@6.8.2` to get around of this type incompatibility issue while we are looking into it?

**Comment by Drillan767** [comment]
Hello @HuiSF, no worries about the inconvenience, I'm just really glad you could reproduce it! I did downgrade to `aws-amplify@6.8.2`, and there's no longer any typescript issues, my project was able to build, thank you!

Do I keep this issue open?

**Comment by ashwinkumar6** [comment]
@Drillan767 the fix is available in [unstable](https://www.npmjs.com/package/aws-amplify/v/6.10.1-unstable.fb8dadb.0), could you please try clearing node_modules and lock file and test on `aws-amplify@unstable` if the issue is resolved

**Comment by ashwinkumar6** [comment]
@Drillan767 
Thanks for raising the issue, this issue is fixed on [aws-amplify@6.10.2](https://www.npmjs.com/package/aws-amplify/v/6.10.2).
